File: server/netpackage.py
# -*- coding: utf-8 -*-
# 网络数据打包工具
# 注意: 打包整形数据时，如果确定整数大小则尽量使用具体的PacketAddInt8，PacketAddInt16，PacketAddInt32，不知道具体大小再使用PacketAddInt
# NOTE: unpack最后记得del netpack对象！！！！！！

from pubdefines import CallManagerFunc

import struct
import pubdefines
import mq
import json

def NetPackagePrepare(byteContent = b""):
	return CNetPackage(byteContent)

class CNetPackage:
	def __init__(self, byteContent):
		self.m_BytesBuffer = byteContent
		self.m_Offset = 0

	def PackInto(self, byteContent):
		self.m_BytesBuffer += byteContent

	def Unpack(self, sType):
		iAddOffset = struct.calcsize(sType)
		byteContent = self.m_BytesBuffer[self.m_Offset:self.m_Offset+iAddOffset]
		self.m_Offset += iAddOffset
		return struct.unpack(sType, byteContent)[0]

	def UnpackEnd(self):
		return self.m_BytesBuffer[self.m_Offset:]

def PacketPrepare(header = 0):
	oNetPack = NetPackagePrepare()
	if header:
		PacketAddInt16(header, oNetPack)
	return oNetPack

def PacketAddInt(iVal, oNetPack):
	if iVal <= 255:
		PacketAddInt8(1, oNetPack)
		PacketAddInt8(iVal, oNetPack)
	elif 255 < iVal <= 65535:
		PacketAddInt8(2, oNetPack)
		PacketAddInt16(iVal, oNetPack)
	elif 65535 < iVal <= 4294967295:
		PacketAddInt8(4, oNetPack)
		PacketAddInt32(iVal, oNetPack)
	elif 4294967296 < iVal <= 9223372036854775807:
		PacketAddInt8(8, oNetPack)
		PacketAddInt64(iVal, oNetPack)
	else:
		PrintError("protocol error: python int value bigger than int64 %s"%(iVal))

def PacketAddInt8(iVal, oNetPack):
	"""
	无符号1字节整形 0-255
	"""
	if not isinstance(iVal, int):
		iVal = int(iVal)
	byteData = struct.pack("B", iVal)
	if byteData:
		oNetPack.PackInto(byteData)

def PacketAddInt16(iVal, oNetPack):
	"""
	无符号2字节整形 0-65535
	"""
	if not isinstance(iVal, int):
		iVal = int(iVal)
	byteData = struct.pack("H", iVal)
	if byteData:
		oNetPack.PackInto(byteData)

def PacketAddInt32(iVal, oNetPack):
	"""
	无符号4字节整形 0-4294967295
	"""
	if not isinstance(iVal, int):
		iVal = int(iVal)
	byteData = struct.pack("I", iVal)
	if byteData:
		oNetPack.PackInto(byteData)

def PacketAddInt64(iVal, oNetPack):
	"""
	无符号8字节整形 0-9223372036854775807 尽量不使用(浪费空间，小程序js只支持到4字节)，长整形可以转成字符串进行压包
	"""
	if not isinstance(iVal, int):
		iVal = int(iVal)
	byteData = struct.pack("Q", iVal)
	if byteData:
		oNetPack.PackInto(byteData)

def PacketAddB(bytes, oNetPack):
	oNetPack.PackInto(bytes)

def PacketAddC(char, oNetPack):
	byteData = struct.pack('c', bytes(char.encode("utf-8")))
	if byteData:
		oNetPack.PackInto(byteData)

def PacketAddS(sVal, oNetPack):
	"""
	默认最长4294967295,8字节的长度最好不要用，需要限制字符串长度
	"""
	sEncodeStr = sVal.encode('utf-8')
	iLen = len(sEncodeStr)
	PacketAddInt(iLen, oNetPack)
	if iLen == 1:
		PacketAddC(sEncodeStr, oNetPack)
	else:
		byteData = struct.pack('%ss' % len(sEncodeStr), sEncodeStr)
		if byteData:
			oNetPack.PackInto(byteData)

def PacketAddBool(bool, oNetPack):
	iVal = 0
	if bool:
		iVal = 1
	PacketAddInt8(iVal, oNetPack)

def PacketAddJSON(data, oNetPack):
	"""
	效率较低，尽量少使用，数据结构复杂时候再用
	"""
	sData = json.dumps(data, ensure_ascii=False)
	PacketAddS(sData, oNetPack)

def PacketSend(link, oNetPack):
	import conf
	if conf.IsGate():
		oMq = mq.GetMq(pubdefines.MSGQUEUE_SEND)
		bData = oNetPack.m_BytesBuffer
		if oMq:
			if oMq.full():
				PrintWarning("net work delay")
				return
			tFlag = CallManagerFunc("link", "GetClientLink", link)
			oMq.put((pubdefines.CLIENT, tFlag, bData))
		del oNetPack
	elif conf.IsGPS():
		import rpc
		import script.player as player
		iServer, iIndex = conf.GetGate()
		iConnectID = player.GetConnectIDByOpenID(link)
		if iConnectID < 0:
			return
		rpc.RemoteCallFunc(iServer, iIndex, None, "netpackage.R_GateSend2Client", iConnectID, oNetPack.m_BytesBuffer)
	else:
		PrintError("the current process can not send data to client!")

def GPSPacketSendByConnectID(iConnectID, oNetPack):
	import rpc
	import conf
	if conf.IsGPS():
		iServer, iIndex = conf.GetGate()
		rpc.RemoteCallFunc(iServer, iIndex, None, "netpackage.R_GateSend2Client", iConnectID, oNetPack.m_BytesBuffer)

def R_GateSend2Client(response, iConnectID, data):
	import conf
	if not conf.IsGate():
		PrintError("the current process can not send data to client direct!")
		return
	oNetPack = PacketPrepare()
	PacketAddB(data, oNetPack)
	PacketSend(iConnectID, oNetPack)
 
def S2SPacketSend(iServer, iIndex, oNetPack):
	oMq = mq.GetMq(pubdefines.MSGQUEUE_SEND)
	bData = oNetPack.m_BytesBuffer
	if oMq:
		if oMq.full():
			PrintWarning("net work delay")
			return
		tFlag = CallManagerFunc("link", "GetLink", iServer, iIndex)
		oMq.put((pubdefines.SERVER, tFlag, bData))
	del oNetPack

'''def PacketSend(iLink, oNetPack):
	oLink = pubdefines.CallManagerFunc("link", "GetLink", iLink)
	if oLink:
		#oLink.m_Socket.transport.write(oNetPack.m_BytesBuffer)
		#oLink.m_Socket.transport.getHandle().sendall(oNetPack.m_BytesBuffer)
	del oNetPack'''

def UnpackPrepare(byteData):
	oNetPackage = NetPackagePrepare(byteData)
	return oNetPackage

def UnpackInt(oNetPackage):
	iByte = UnpackInt8(oNetPackage)
	if iByte == 1:
		return UnpackInt8(oNetPackage)
	elif iByte == 2:
		return UnpackInt16(oNetPackage)
	elif iByte == 4:
		return UnpackInt32(oNetPackage)
	elif iByte == 8:
		return UnpackInt64(oNetPackage)
	else:
		PrintError("netpackage error: unpack int bigger than int64")

def UnpackInt8(oNetPackage):
	"""
	无符号1字节整形 0-255
	"""
	return int(oNetPackage.Unpack("B"))

def UnpackInt16(oNetPackage):
	"""
	无符号2字节整形 0-65535
	"""
	return int(oNetPackage.Unpack("H"))

def UnpackInt32(oNetPackage):
	"""
	无符号4字节整形 0-4294967295
	"""
	return int(oNetPackage.Unpack("I"))

def UnpackInt64(oNetPackage):
	"""
	无符号8字节整形 0-9223372036854775807 尽量不使用，长整形可以转成字符串进行压包
	"""
	return int(oNetPackage.Unpack("Q"))

def UnpackBool(oNetPackage):
	iVal = UnpackInt8(oNetPackage)
	bVal = False
	if iVal:
		bVal = True
	return bVal

def UnpackC(oNetPackage):
	return oNetPackage.Unpack("c").decode("utf-8")

def UnpackEnd(oNetPackage):
	return oNetPackage.UnpackEnd()

def UnpackS(oNetPackage):
	"""
	默认最长4294967295字节，需要限制字符串长度
	"""
	iLen = UnpackInt(oNetPackage)
	if iLen == 1:
		return UnpackC(oNetPackage)
	else:
		return oNetPackage.Unpack("%ss" % iLen).decode("utf-8")

def UnpackJSON(oNetPackage):
	"""
	效率较低，尽量少使用，数据结构复杂时候再用
	"""
	sData = UnpackS(oNetPackage)
	return json.loads(sData)


"""
Format	C Type				Python type			Standard size	Notes
x		pad byte			no value	 	 
c		char				string of length 1	1	 
b		signed char			integer				1				(3)
B		unsigned char		integer				1				(3)
?		_Bool				bool				1				(1)
h		short				integer				2				(3)
H		unsigned short		integer				2				(3)
i		int					integer				4				(3)
I		unsigned int		integer				4				(3)
l		long				integer				4				(3)
L		unsigned long		integer				4				(3)
q		long long			integer				8				(2), (3)
Q		unsigned long long	integer				8				(2), (3)
f		float				float				4				(4)
d		double				float				8				(4)
s		char[]				string				1	 
p		char[]				string	 	 
P		void *				integer	 							(5), (3)

str.encode(‘utf-8')
bytes.decode(‘utf-8')
"""

File: client/utils/netpackage.js
/*
网络数据打包工具
注意: 打包整形数据时，如果确定整数大小则尽量使用具体的PacketAddInt8，PacketAddInt16，PacketAddInt32，不知道具体大小再使用PacketAddInt
*/

import {ARRAYBUFFER_UTILS, STRING_ARRAYBUFFER} from 'handlebytes.js'
import {execServerCommand} from 'servercommand.js'

let PROTOCOL_CB
if (typeof(PROTOCOL_CB) == "undefined"){
  PROTOCOL_CB = {}
}

function removeProtocolCB(iProtocol){
  if (iProtocol in PROTOCOL_CB){
    let cbFunc = PROTOCOL_CB[iProtocol]
    let state = -1 //表示超时
    cbFunc(state, null)
    delete PROTOCOL_CB[iProtocol]
  }
}

export function netCommand(header, oNetPack){
  if (! (header in PROTOCOL_CB)){
    execServerCommand(header, oNetPack)
    return
  }
  let cbFunc = PROTOCOL_CB[header]
  delete PROTOCOL_CB[header]
  try{
    let state = 1 //表示回调成功
    cbFunc(state, oNetPack)
  } catch(err){
    console.error(err)
  }
}

class CNetPackage{
  constructor(arraybuffer){
    this.m_Offset = 0
    this.m_DataList = []
    this.m_ArrayBuffer = arraybuffer
    this.m_DataView = null
  }

  PackInto(sType, data){
    this.m_DataList.push(data)
  }

  PackAll(){
    let tmpBuffer = ARRAYBUFFER_UTILS.concact(this.m_DataList)
    this.m_ArrayBuffer = tmpBuffer
  }

  UnpackPrepare(){
    this.m_DataView = new DataView(this.m_ArrayBuffer)
  }

  Unpack(sType, iStrLen = 0){
    switch (sType){
      case "B":
        let data1 = this.m_DataView.getUint8(this.m_Offset, true)
        this.m_Offset += 1
        return data1
      case "H":
        let data2 = this.m_DataView.getUint16(this.m_Offset, true)
        this.m_Offset += 2
        return data2
      case "I":
        let data3 = this.m_DataView.getUint32(this.m_Offset, true)
        this.m_Offset += 4
        return data3
      case "S":
        let data4 = this.UnpackS(iStrLen)
        this.m_Offset += iStrLen
        return data4
    }
  }

  UnpackS(iStrLen){
    return STRING_ARRAYBUFFER.dataview2string(this.m_DataView, this.m_Offset, iStrLen)
  }
}

export const NetPack = {
  PacketPrepare: function PacketPrepare (header, cbFunc=null) {
    if (cbFunc){
      PROTOCOL_CB[header] = cbFunc
    }
    setTimeout(removeProtocolCB, 10000, header)
    let oNetPack = new CNetPackage(null)
    this.PacketAddInt16(header, oNetPack)
    return oNetPack
  },

  PacketAddInt: function PacketAddInt(iVal, oNetPack){
    if (iVal <= 255){
      this.PacketAddInt8(1, oNetPack)
      this.PacketAddInt8(iVal, oNetPack)
    }
    else if (iVal > 255 && iVal <= 65535 ) {
      this.PacketAddInt8(2, oNetPack)
      this.PacketAddInt16(iVal, oNetPack)
    }
    else if (iVal > 65535 && iVal <= 4294967295 ) {
      this.PacketAddInt8(4, oNetPack)
      this.PacketAddInt32(iVal, oNetPack)
    }
    else{
      throw new Error("val pack exceeded", iVal)
    }
  },

  PacketAddInt8: function PacketAddInt8(iVal, oNetPack){
    let uint8array = new Uint8Array(1)
    uint8array[0] = iVal
    oNetPack.PackInto("B", uint8array.buffer)
  },

  PacketAddInt16: function PacketAddInt16(iVal, oNetPack){
    let uint16array = new Uint16Array(1)
    uint16array[0] = iVal
    oNetPack.PackInto("H", uint16array.buffer)
  },

  PacketAddInt32: function PacketAddInt32(iVal, oNetPack){
    let uint32array = new Uint32Array(1)
    uint32array[0] = iVal
    oNetPack.PackInto("I", uint32array.buffer)
  },

  PacketAddBool: function PacketAddBool(bVal, oNetPack){
    let iVal = 0
    if (bVal == true){
      iVal = 1
    }
    this.PacketAddInt8(iVal, oNetPack)
  },

  PacketAddS: function PacketAddS(str, oNetPack){
    let arraybuffer = STRING_ARRAYBUFFER.string2arraybuffer(str)
    let iStrLen = arraybuffer.byteLength
    this.PacketAddInt(iStrLen, oNetPack)
    oNetPack.PackInto("S", arraybuffer)
  },

  PacketSend: function PacketSend(oNetPack){
    let app = getApp()
    oNetPack.PackAll()
    if (!app.getConnectState()) {
      console.log("[error] disconnect with server, send data failed")
      return
    }
    let conn = app.getTcpConnect()
    conn.write(oNetPack.m_ArrayBuffer)
  },

  UnpackPrepare: function UnpackPrepare(arraybuffer){
    let oNetPack = new CNetPackage(arraybuffer)
    oNetPack.UnpackPrepare()
    return oNetPack
  },

  UnpackInt: function UnpackInt(oNetPack){
    let iByte = this.UnpackInt8(oNetPack)
    if (iByte == 1){
      return this.UnpackInt8(oNetPack)
    }
    else if (iByte == 2){
      return this.UnpackInt16(oNetPack)
    }
    else if (iByte == 4){
      return this.UnpackInt32(oNetPack)
    }
    else{
      throw new Error("netpackage error: UnpackInt bigger than int32")
    }
  },

  UnpackInt8: function UnpackInt8(oNetPack){
    return oNetPack.Unpack("B")
  },

  UnpackInt16: function UnpackInt16(oNetPack){
    return oNetPack.Unpack("H")
  },

  UnpackInt32: function UnpackInt32(oNetPack){
    return oNetPack.Unpack("I")
  },

  UnpackBool: function UnpackBool(oNetPack){
    let iVal = this.UnpackInt8(oNetPack)
    let bVal = false
    if (iVal == 1){
      bVal = true
    }
    return bVal
  },

  UnpackS: function UnpackS(oNetPack){
    let iLen = this.UnpackInt(oNetPack)
    return oNetPack.Unpack("S", iLen)
  }
}